rsconnect/utils_package.py

```diff
@@ -0,0 +1,265 @@
+    modified requirements.txt.
+
+    Note that package_name is a regular expression, so if the target package name
+    contains a ".", it should be escaped, as in "foo\\.bar".
+    """
+    lines = requirements_txt.split("\n")
+    new_lines: list[str] = []
+
+    for line in lines:
+        new_line = re.sub(f"^{package_name}([^a-zA-Z0-9._-].*?)?$", replacement, line)
+        new_lines.append(new_line)
+
+    return "\n".join(new_lines)
+
+
+def fix_starlette_requirements(
+    environment: Environment,
+    app_mode: AppMode,
+    connect_version_string: str,
+) -> Environment:
+    """
+    Ensure that the starlette version in an Environment is compatible with the Connect
+    server.
+
+    If the app mode is PYTHON_SHINY and the Connect server version is less than
+    2024.01.0, then make sure the starlette version is less than 0.35.0, due to an
+    incompatibility with between older Connect<2024.01.0 and starlette>=0.35.0.
+
+    After all users are on Connect 2024.01.0 or later, this function can be removed.
+
+    For more information, see https://github.com/posit-dev/py-shiny/issues/1114
+
+    :return: Either a modified environment (if changes were needed), or the original
+    environment.
+    """
+    if not (app_mode == AppModes.PYTHON_SHINY and compare_semvers(connect_version_string, "2024.01.0") == -1):
+        return environment
+
+    requirements_txt = cast(str, environment.contents)
+
+    reqs = parse_requirements_txt(requirements_txt)
+    starlette_req = find_package_info("starlette", reqs)
+
+    # If didn't ask for specific version of starlette, add starlette<0.35.0.
+    # If pinned version is <0.35.0, do nothing.
+    # If pinned version is >=0.35.0, change the version and emit warning
+    # If more complex version spec (that doesn't use ==), do nothing.
+
+    if starlette_req is None:
+        # starlette is not listed in requirements.
+        # Add starlette<0.35.0 to requirements.
+        requirements_txt = requirements_txt.rstrip("\n") + "\nstarlette<0.35.0\n"
+        environment = environment._replace(contents=requirements_txt)
+
+    elif len(starlette_req.specs) == 0:
+        # starlette is in requirements, but without a version specification.
+        # Replace it with starlette<0.35.0.
+        requirements_txt = replace_requirement("starlette", "starlette<0.35.0", requirements_txt)
+        environment = environment._replace(contents=requirements_txt)
+
+    elif len(starlette_req.specs) == 1 and starlette_req.specs[0].operator == "==":
+        if compare_semvers(starlette_req.specs[0].version, "0.35.0") >= 0:
+            # starlette is in requirements.txt, but with a version spec that is
+            # not compatible with this version of Connect.
+            print(
+                "Starlette version is 0.35.0 or greater, but this version of Connect "
+                "requires starlette<0.35.0. Setting to <0.35.0."
+            )
```

Comment:
This uses `print()` to emit a message for the user. Is that the right way to print messages?

Reply:
Elsewhere we are using:
```
from .log import logger
logger.warn(...)
```